File: app/src/main/java/com/ternaryop/photoshelf/adapter/feedly/FeedlyContentDelegate.java
package com.ternaryop.photoshelf.adapter.feedly;

import java.net.URI;
import java.net.URISyntaxException;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.format.DateUtils;

import com.ternaryop.feedly.FeedlyContent;
import com.ternaryop.photoshelf.R;
import com.ternaryop.utils.DateTimeUtils;

/**
 * Contains fields related to UI state
 */
public class FeedlyContentDelegate extends FeedlyContent {
    private final FeedlyContent delegated;
    public boolean checked = true;
    public long lastPublishTimestamp;
    private String domain;

    public FeedlyContentDelegate(FeedlyContent delegated) {
        this.delegated = delegated;
        try {
            domain = new URI(getOriginId()).getHost();
        } catch (URISyntaxException ignored) {
        }
    }

    @Override
    public String getId() {
        return delegated.getId();
    }

    @Override
    public String getTitle() {
        return delegated.getTitle();
    }

    @Override
    public String getOriginId() {
        return delegated.getOriginId();
    }

    @Override
    public long getActionTimestamp() {
        return delegated.getActionTimestamp();
    }

    @Override
    public void setId(String id) {
        delegated.setId(id);
    }

    @Override
    public void setTitle(String title) {
        delegated.setTitle(title);
    }

    @Override
    public void setOriginId(String originId) {
        delegated.setOriginId(originId);
    }

    @Override
    public void setActionTimestamp(long actionTimestamp) {
        delegated.setActionTimestamp(actionTimestamp);
    }

    @Override
    public Origin getOrigin() {
        return delegated.getOrigin();
    }

    @Override
    public void setOrigin(Origin origin) {
        delegated.setOrigin(origin);
    }

    @NonNull
    public String getLastPublishTimestampAsString(Context context) {
        if (lastPublishTimestamp <= 0) {
            return context.getString(R.string.never_published);
        }
        return DateTimeUtils.formatPublishDaysAgo(lastPublishTimestamp * 1000, DateTimeUtils.APPEND_DATE_FOR_PAST_AND_PRESENT);
    }

    @NonNull
    public String getActionTimestampAsString(Context context) {
        return DateUtils.getRelativeTimeSpanString(context, getActionTimestamp()).toString();
    }

    public String getDomain() {
        return domain;
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/adapter/feedly/FeedlyContentDelegate.kt
package com.ternaryop.photoshelf.adapter.feedly

import android.content.Context
import android.text.format.DateUtils
import android.text.format.DateUtils.SECOND_IN_MILLIS
import com.ternaryop.feedly.FeedlyContent
import com.ternaryop.photoshelf.R
import com.ternaryop.utils.date.APPEND_DATE_FOR_PAST_AND_PRESENT
import com.ternaryop.utils.date.formatPublishDaysAgo
import java.net.URI
import java.net.URISyntaxException

/**
 * Convert Collection<FeedlyContent> to List<FeedlyContentDelegate>
 */
fun Collection<FeedlyContent>.toContentDelegate() = map { FeedlyContentDelegate(it) }

fun Collection<FeedlyContentDelegate>.titles(): Collection<String> {
    val titles = HashSet<String>(size)
    for (fc in this) {
        // replace any no no-break space with whitespace
        // see http://www.regular-expressions.info/unicode.html for \p{Zs}
        titles.add(fc.title.replace("""\p{Zs}""".toRegex(), " "))
    }
    return titles
}

/**
 * update lastPublishTimestamp and the tag fields
 */
fun Collection<FeedlyContentDelegate>.updateLastPublishTimestamp(tagPairList: Map<String, Long>): Collection<FeedlyContentDelegate> {
    for (fc in this) {
        updateLastPublishTimestamp(fc, tagPairList)
    }
    return this
}

private fun updateLastPublishTimestamp(fc: FeedlyContentDelegate, tagPairList: Map<String, Long>) {
    val title = fc.title
    for ((tag, time) in tagPairList) {
        if (tag.regionMatches(0, title, 0, tag.length, ignoreCase = true)) {
            fc.lastPublishTimestamp = time
            fc.tag = tag
        }
    }
}

/**
 * Contains fields related to UI state
 */
class FeedlyContentDelegate(private val delegated: FeedlyContent) : FeedlyContent by delegated {
    var isChecked = true
    var lastPublishTimestamp = Long.MIN_VALUE
    var domain: String? = null
        private set
    var tag: String? = null

    init {
        try {
            domain = URI(originId).host
        } catch (ignored: URISyntaxException) {
        }
    }

    fun getLastPublishTimestampAsString(context: Context): String {
        return if (lastPublishTimestamp <= 0) {
            context.getString(R.string.never_published)
        } else (lastPublishTimestamp * SECOND_IN_MILLIS).formatPublishDaysAgo(APPEND_DATE_FOR_PAST_AND_PRESENT)
    }

    fun getActionTimestampAsString(context: Context): String {
        return DateUtils.getRelativeTimeSpanString(context, actionTimestamp).toString()
    }

    fun compareTitle(other: FeedlyContent): Int = title.compareTo(other.title, ignoreCase = true)

    fun compareActionTimestamp(other: FeedlyContent): Int {
        val at1 = actionTimestamp
        val at2 = other.actionTimestamp
        return if (at1 == at2) 0 else if (at1 < at2) 1 else -1
    }

    fun compareLastTimestamp(other: FeedlyContentDelegate): Int {
        val t1 = lastPublishTimestamp
        val t2 = other.lastPublishTimestamp

        return if (t1 == t2) {
            compareTitle(other)
        } else {
            if (t1 < t2) -1 else 1
        }
    }
}
